agentspawner/daemon.py

```diff
@@ -0,0 +1,126 @@
+
+from resalloc.client import (
+    Connection as ResallocConnection,
+    Ticket,
+)
+
+RESALLOC_SERVER = "http://localhost:49100"
+
+
+class SpawnerPool:
+    """
+    Manage ideal number of "agent" like resources in Resalloc.
+    """
+    sleep = 30
+
+    def __init__(self, resalloc_connection, logger):
+        self.tags = ["A"]
+        # TODO: use a persistent storage so we can restart the process
+        self.tickets = []
+        self.conn = resalloc_connection
+        self.log = logger
+
+    def call_converge_to(self):
+        """ Execute the configured hook script """
+        while True:
+            result = subprocess.run(["./hook-converge-to"], capture_output=True,
+                                    check=False)
+            if result.returncode == 0:
+                try:
+                    return int(result.stdout.decode("utf-8").strip())
+                except ValueError:
+                    pass
+
+            self.log.debug("Failing to run converge-to hook")
+
+    def call_take(self, data):
+        """
+        Call hook that prepares the resource
+        """
+        return not subprocess.run(["./hook-take", f"{data}"], check=True)
+
+    def call_release(self, data):
+        """
+        Call hook that releases the resource
+        """
+        result = subprocess.run(["./hook-release", f"{data}"], check=False)
+        return not result.returncode
+
+    def start(self, count):
+        """ Start N agent-like resources """
+        self.log.info("Starting %s resources", count)
+        for _ in range(count):
+            ticket = self.conn.newTicket(self.tags)
+            self.log.debug("Taking ticket id %s", ticket.id)
+            self.tickets.append(ticket.id)
+            data = ticket.wait()
+            self.call_take(data)
+
+    def try_to_stop(self, to_stop):
+        """
+        Attempt to stop TO_STOP resources by closing Resalloc tickets.  Not all
+        the resources may be closed at this time.
+        """
+        self.log.info("Trying to stop %s resources", to_stop)
+        stopped = 0
+        for ticket_id in self.tickets:
+            if stopped >= to_stop:
+                break
+
+            ticket = Ticket(ticket_id, connection=self.conn)
+            data = ticket.collect()
+            if not self.call_release(data):
+                self.log.debug("Can't release %s", ticket.id)
+                continue
+
+            self.log.debug("Closing ticket %s", ticket.id)
+            ticket.close()
+            self.tickets.remove(ticket_id)
+            stopped += 1
+
+
+    def loop(self):
+        """
+        Periodically query the ideal number of builders, and attempt to converge
+        to the ideal state.
+        """
+        while True:
+            start = time.time()
+            todo = self.call_converge_to() - len(self.tickets)
+            if todo > 0:
+                self.start(todo)
+            elif todo < 0:
+                self.try_to_stop(-todo)
```

Comment:
There should be a periodic check here to find virtual machines that can not be used anymore and should be deleted.

Reply:
Can you elaborate on this one?   The `try_to_stop` hook actually does this?

Reply:
`try_to_stop` hook tries to close tickets based on a number, so if the number do not change, the tickets would not close. We would like to have throwaway workers in openscanhub, that is, a worker should be deleted once a task assigned to it has been completed. This would require querying the database and it should be done through a separate hook.

On a sidenote, we should regularly perform health checks on the workers and if a worker ends up in an unusable state, its related ticket should be closed.

Reply:
Good point. Discussed on Slack 1:1, the ticket needs to be closed for finished agents .... because agents can not close the ticket themselves.  We could rely on "FAILED" ticket state (if the VM/resource goes down by itself) but that's not ideal for OSH.

Reply:
If an agent goes down by itself and the ticket is in failed state, there should be a separate hook for it to remove the agent from osh (kobo) database.